Modular Population of Topics, Sections, and Resources with New Database System

programnoir opened the pull request:
This references the following issues:
* #2 Populate Sections Modularly and Responsively
* #3 Remove jQuery Dependency from search.js
* New bug located: Z-index for header not properly set.

As the title indicates, db.js has been split into two files, one for initializing all the values, and one for accessing its members. See issue 2 for details about why we're transitioning to this system. This is essentially a replacement of the support beam to the house. StudyGameDev was very stable, before, but it can overwhelm the platform by loading thousands of nodes upfront, which would quickly eat up lots of RAM if it were to continue to grow. Not only that, but the site depended on tons of JavaScript operations tied to DOM additions, which itself makes an impact on score.

The pros here are many; decent savings on RAM, the database is able to expand again, jQuery can be removed, and lots of other small details.

However, this current pull request has a number of glaring problems.

First off, referring now to #5, the code here is atrocious, and the function design is also pretty terrible-looking. StudyGameDev was never the most engineered design, but any notion of a consistent variable naming scheme for instance has been trashed. Code documentation is very lacking, and overall this thing is ugly as sin.

The second problem is that due to the removal of all of those nodes in `index.php`, the code generator (gencode.js) no longer works. This is of course a fairly easy fix by simply modifying it to iterate through db_topics for the values, it just needs its own issue and branch to fix.

The third problem is that this is a pull request with known bugs. I'm a little new to having multiple asynchronous calls going on and while they are very quick, if you time the wrong button press just when some population is going on, you can run into some major catastrophes. In short - the asynchronous population scripts do not have enough safety and cancellation measures, although there are some in place, such as, I think, the search term population scripts. Don't quote me on that one, though.

This is being published as a pull request for the sake of transparency and discussion. I'd like to take discussion about the coding style practices to the topic on #5. Thank you.

programnoir committed:
More descriptive variables

Adhering to a better code format standard and giving more descriptive variable names. Mainly making this commit because I stupidly tried to work on master lol

programnoir committed:
Including TaffyDB

Including TaffyDB and referring to the license in the readme.md file.

programnoir committed:
Database transferred to Taffy

TaffyDB is now the host of the resource, rather than using an array.

programnoir committed:
WIP: Menu Modular Population

The menu will modularly populate the items (not yet, menu needs to access this, first)

programnoir committed:
Menus modularly populate

Menus now modularly populate, saving on the number of nodes in the page at a time.

programnoir committed:
Modular population of topics

Reduced number of DOM nodes by about half, removing and adding event listeners, and in general breaking every other feature of the site.

programnoir committed:
Search function: Empty restores section

When you enter a character into the search field, it waits about two seconds to conduct a search. When it is empty, it immediately reverts back to the section you were previously exploring.

programnoir committed:
Search initiates a query and stores it in refs

With a few bug fixes for cancelling the operation and an actual function for placing everything in as planned, we should be good to go.

programnoir committed:
Finally populated search terms

It's a broken disaster, but it is happening.

programnoir committed:
CSS fix and topic clearance fix

Transitioning from search results to menu clicks is fixed. I think.

programnoir committed:
From menu click to search fixed

Searches now properly populate and depopulate with respect to menu clicks.

programnoir committed:
jQuery Removal & Readme Update

Updates to readme.md and jquery inclusion that is no longer used.

nitronova commented:
I'm gonna go ahead and attach a meme here. Beware ye who enter.
![meme](https://user-images.githubusercontent.com/1452074/58379528-32612900-7f6a-11e9-8d1d-b7cefb7e3963.png)

programnoir committed:
Attempts to reduce memory leaks?

Apparently there are memory leaks associated with the DOM tree but I can't get it to fully revert.

programnoir committed:
More RAM control

Figured maybe I was hitting the end of an array before I removed all of the elements, so theoretically I should iterate in the opposite direction.

programnoir committed:
let keyword for scope

mainly a minor scope keyword change

programnoir committed:
clear object from variable

a little more RAM control

programnoir committed:
Renaming Classes/IDs, style adherence in index.php

Renames all classes and IDs in index.php across all files. index.php got a massive style overhaul to ensure that tabbing and attribute order was right.

programnoir committed:
Further style definition and adherence

Removed need for topics to append/remove an unused "show" tag. Resource tag classes are now properly named to be descriptive.

nitronova commented:
I think I would consider this "a bad pull request" as it is trying to fix too many issues at once. But because the project lacks decoupling, this is impossible to avoid.

programnoir committed:
Style adherence progress

Class redefinitions and reorganization of style rules.

programnoir committed:
More style.css style adherence

More parentheses spacing and rule reorganization

programnoir committed:
Variable/function renaming

Renaming variable and function names in an attempt to add more clarity as to what the scripts are doing.

programnoir committed:
Style Adherence in awesome-menu.js

awesome-menu.js gets new function and variable names, mainly. Some corrections in other files to address accidentally created bugs are applied.

programnoir committed:
CodeGen and Awesome: Style adherance

Renaming, more updates to let Code Gen work with the new db, etc.

programnoir committed:
Search.js style adherence

Final style adherence commit.

programnoir committed:
Remove annoying comments

No longer to-do

nitronova commented:
Whew. Okay so, everything has been renamed with as much consistency to a style adherence as possible. Names are definitely better, for the most part, so the code is at least a little bit more coherent-looking.

nitronova commented:
Merging now, because ain't nobody gonna review this trash lol.

From now on, my pull requests solve one (1) issue at a time.
